src/gfx/convert.cpp

```diff
@@ -0,0 +1,1001 @@
+			if (color->isTransparent()) {
+				continue;
+			}
+			if (!color->isGray()) {
+				options.verbosePrint(Options::VERB_DEBUG,
+				                     "Found non-gray color #%08x, not using grayscale sorting\n",
+				                     color->toCSS());
+				return false;
+			}
+			uint8_t mask = 1 << color->grayIndex();
+			if (bins & mask) { // Two in the same bin!
+				options.verbosePrint(
+				    Options::VERB_DEBUG,
+				    "Color #%08x conflicts with another one, not using grayscale sorting\n",
+				    color->toCSS());
+				return false;
+			}
+			bins |= mask;
+		}
+		return true;
+	}
+
+	/**
+	 * Reads a PNG and notes all of its colors
+	 *
+	 * This code is more complicated than strictly necessary, but that's because of the API
+	 * being used: the "high-level" interface doesn't provide all the transformations we need,
+	 * so we use the "lower-level" one instead.
+	 * We also use that occasion to only read the PNG one line at a time, since we store all of
+	 * the pixel data in `pixels`, which saves on memory allocations.
+	 */
+	explicit Png(std::string const &filePath) : path(filePath), colors() {
+		if (file.open(path, std::ios_base::in | std::ios_base::binary) == nullptr) {
+			fatal("Failed to open input image (\"%s\"): %s", path.c_str(), strerror(errno));
+		}
+
+		options.verbosePrint(Options::VERB_LOG_ACT, "Opened input file\n");
+
+		std::array<unsigned char, 8> pngHeader;
+
+		if (file.sgetn(reinterpret_cast<char *>(pngHeader.data()), pngHeader.size())
+		        != static_cast<std::streamsize>(pngHeader.size()) // Not enough bytes?
+		    || png_sig_cmp(pngHeader.data(), 0, pngHeader.size()) != 0) {
+			fatal("Input file (\"%s\") is not a PNG image!", path.c_str());
+		}
+
+		options.verbosePrint(Options::VERB_INTERM, "PNG header signature is OK\n");
+
+		png = png_create_read_struct(PNG_LIBPNG_VER_STRING, (png_voidp)this, handleError,
+		                             handleWarning);
+		if (!png) {
+			fatal("Failed to allocate PNG structure: %s", strerror(errno));
+		}
+
+		info = png_create_info_struct(png);
+		if (!info) {
+			png_destroy_read_struct(&png, nullptr, nullptr);
+			fatal("Failed to allocate PNG info structure: %s", strerror(errno));
+		}
+
+		png_set_read_fn(png, this, readData);
+		png_set_sig_bytes(png, pngHeader.size());
+
+		// TODO: png_set_crc_action(png, PNG_CRC_ERROR_QUIT, PNG_CRC_WARN_DISCARD);
+
+		// Skipping chunks we don't use should improve performance
+		// TODO: png_set_keep_unknown_chunks(png, ...);
+
+		// Process all chunks up to but not including the image data
+		png_read_info(png, info);
+
+		int bitDepth, interlaceType; //, compressionType, filterMethod;
+
+		png_get_IHDR(png, info, &width, &height, &bitDepth, &colorType, &interlaceType, nullptr,
+		             nullptr);
+
+		if (width % 8 != 0)
+			fatal("Image width (%" PRIu32 " pixels) is not a multiple of 8!", width);
+		if (height % 8 != 0)
+			fatal("Image height (%" PRIu32 " pixels) is not a multiple of 8!", height);
```

Comment:
You commented below that we should always use braces; this and some other block in this function do not.

Reply:
And just like that, this file has just reached a thousand lines! :D